product_variant_exclusion/models/product.py

```diff
@@ -0,0 +1,174 @@
+                                         compute='_compute_allowed_values')
+
+    @api.depends('attribute_line_ids.value_ids')
+    def _compute_allowed_values(self):
+        for template in self:
+            template.allowed_value_ids = template.attribute_line_ids.mapped(
+                'value_ids')
+
+    @api.multi
+    def write(self, vals):
+        res = super(ProductTemplate, self).write(vals)
+        if 'variant_exclusion_ids' in vals:
+            self.create_variant_ids()
+            self.delete_variant_ids()
+        return res
+
+    @api.multi
+    def delete_variant_ids(self):
+        exclusions = self.env['product.variant.exclusion'].search(
+            [('product_tmpl_id', '=', self.id)])
+        for template in self:
+            for product in template.product_variant_ids:
+                for excl in exclusions:
+                    if set(excl.attribute_value_ids.ids) <= set(
+                            product.attribute_value_ids.ids):
+                        try:
+                            with self.env.cr.savepoint():
+                                product.unlink()
+                        except (psycopg2.Error, ValidationError):
+                            product.write({'active': False})
+                        break
+
+
+class ProductAttributeLine(models.Model):
+
+    _inherit = 'product.attribute.line'
+
+    @api.multi
+    def write(self, vals):
+        attr_values = vals.get('value_ids', False) and vals.get(
+            'value_ids')[0][2]
+        if attr_values and len(attr_values) <= 1:
+            for line in self:
+                for exclusion in line.product_tmpl_id.variant_exclusion_ids:
+                    if set(attr_values) <= set(
+                            exclusion.attribute_value_ids.ids):
+                        raise UserError(_(
+                            'You cannot leave only one attribute value on a '
+                            'line if this value is part of an exclusion'))
+        if self.value_ids.ids != attr_values:
+            new_values = self.env[
+                'product.attribute.value'].browse(attr_values)
+            exclusion_values = self.product_tmpl_id.variant_exclusion_ids\
+                .mapped('attribute_value_ids')
+            for new_val in new_values:
+                if new_val in exclusion_values:
+                    raise UserError(_('You cannot remove an attribute value '
+                                      'from an attribute line, if this value '
+                                      'is part of an exclusion.'))
+        return super(ProductAttributeLine, self).write(vals)
+
+    @api.multi
+    def unlink(self):
+        for line in self:
+            for val_id in line.value_ids.ids:
+                if val_id in line.product_tmpl_id.variant_exclusion_ids.mapped(
+                        'attribute_value_ids').ids:
+                    raise UserError(_(
+                        'You cannot delete an attribute line when '
```

Comment:
Again why not?  Just unlink the exclusions as well couldn't you?  OK maybe in this one there is an issue if its a 3+ way exclusion, because you don't know if the equivalent 2 way should now be excluded.

I'm OK with all 3 errors on reflection, its not like people delete attribute lines that often.